toptal/picasso
# chore: bundle Picasso as CommonJS

No ticket (yet)

### Description

[Seeing that it was a dream for the team's developers](https://github.com/toptal/talent-portal-frontend/pull/546#issuecomment-607072591), I took the initiative and have achieved good results. It's working properly with tree-shaking.

Benefits of exporting as CommonJS is not needing to transpile stuff from `node_modules` folder, so builds will be faster. Tree-shaking was achieved exporting both ES Modules and CommonJS variations, a common pattern which is widely used (`date-fns` and `material-ui` are good examples).

### How to test

Run `yarn build:dist` on this branch in Picasso's repository. After the command is complete, the generated packages will be grouped at `packages/build` folder.

Checkout on the `commonjs` branch on `talent-portal-frontend` repository. Run `yarn install`. After running, copy the contents of `picasso/packages/build` folder to `talent-portal-frontend/node_modules/@toptal`, replacing the folders. Before pasting on `node_modules/@toptal`, delete the old folders, just in case...

It should run the app normally. Run the build process and you should see the build size not differ.

### Screenshots

Building `talent-portal-frontend` from  `master`

![image](https://user-images.githubusercontent.com/26530524/78197731-9c850d00-745c-11ea-8bcc-e74bd40e62f5.png)

Building `talent-portal-frontend` from `commonjs`

![image](https://user-images.githubusercontent.com/26530524/78197710-8c6d2d80-745c-11ea-8af1-7c963bd6ee8a.png)

### What's missing

- Approval from you guys;
- Fix Storybook and tests, they're probably broken;
- Document the migration for the new packages (`picasso-utils` and `picasso-icons`). It could be that we create a code mod to migrate automatically.

### Why have you separated `utils` and `icons` to new packages???!!?!?!?

Simply because I think we have a better separation of concerns with it.

```
import { Container } from '@toptal/picasso'
import { Check16 } from '@toptal/picasso-icons'
```

Looks better. I can see if it's possible to revert if you don't like it, but for me it's a yes.
Also it's needed to confirm if CommonJS exports works properly with nested paths (`@toptal/picasso/Icon`). I believe not, but believing doesn't work for computer scientists, so testing would be necessary.
